src/pkg/packager/common.go

```diff
@@ -440,6 +443,75 @@ func (p *Packager) validatePackageChecksums() error {
 	return nil
 }
 
+// validatePackageArchitecture validates that the package architecture matches the target cluster architecture.
+func (p *Packager) validatePackageArchitecture() error {
+	var (
+		applianceMode bool
+		clusterArch   string
+		err           error
+		k8sTarget     bool
+	)
+
+	// Iterate over the package components.
+	components := p.getValidComponents()
+	for _, component := range components {
+		// Determine whether we are working with an appliance mode init package.
+		if component.Name == k8s.DistroIsK3s && p.cfg.Pkg.Kind == "ZarfInitConfig" {
+			applianceMode = true
+		}
+
+		// Determine whether we are deploying k8s resources.
+		if component.Images != nil {
+			k8sTarget = true
+		}
+	}
+
+	// If we're working with an init package deploying k3s(appliance mode),
+	// set the clusterArch to the machine we're running on.
+	if applianceMode {
+		clusterArch = runtime.GOARCH
+
+		if p.arch != clusterArch {
+			return fmt.Errorf(lang.CmdPackageDeployValidateArchitectureErr, p.arch, clusterArch)
+		}
+
+		return nil
+	}
```

Comment:
This should be moved to the `k3s` component of the init package as an `onDeploy` action

Reply:
Do you see this being a `zarf internal` command, or using a shell script?

Reply:
Probably a shell script in the action - because `k3s` is linux only you could most likely use `uname -r` or `arch`.  Note that that won't quite match the goarch through so there may be some additional checking needed.

Reply:
Changes made. Will do some experimenting with using an `onDeploy` action

Reply:
@Racer159 the error message when `zarf init` fails doesn't clearly describe why. Is there a way to ensure the error message is clear as to why the deployment failed due to architecture mismatch when having this logic in the `onDeploy` action section of a Zarf init package? https://github.com/defenseunicorns/zarf/actions/runs/4854820097/jobs/8652653901?pr=1495#step:6:4473

Reply:
For further clarification, I already tried to `echo` a custom error message, but it doesn't get reported out on failure from the `onDeploy` action.

Reply:
Would something like this work?

```
        before:
          - cmd: "if [ \"hello\" != \"world\" ]; then echo \"I failed\" && exit 1; fi"
```

Reply:
https://github.com/defenseunicorns/zarf/actions/runs/4855678706/jobs/8654471359?pr=1495#step:6:4467

Looks like adding an echo statement doesn't do anything. The error gets reported as failing on `systemctl start k3s`.

Maybe checking for exit codes other than 0 when running actions and exiting immediately would fix this?